### PR TITLE
Introduce socket plugin helper

### PR DESCRIPTION
**Pull request by tagomoris**
This change introduces socket plugin helper, and migrate `out_forward` to v0.14 API fully with plugin helpers.

This change does:
* add `fluent/plugin_helper/socket` and `#socket_create` (`_tcp`, `_udp`) methods
* fix server plugin helper to call it with existing socket (only for UDP, to create a server without any assigned port)
* fix output plugin to move it into recovery (retry) state when chunk is rollbacked in async mode
* rewrite `out_forward` with socket/timer/server/thread plugin helpers using `#try_write` for at-least-once mode

This change doesn't:
* add tests for socket plugin helper itself ... not yet
* tcp connection pools and/or keepalive implementation

### CODE REVIEW

**Comment by tagomoris** [comment]
@repeatedly Could you review this change?

File: lib/fluent/plugin/output.rb [hunk 1]
```diff
@@ -898,9 +900,9 @@ def commit_write(chunk_id, delayed: @delayed_commit, secondary: false)
         @retry_mutex.synchronize do
           if @retry # success to flush chunks in retries
             if secondary
-              log.warn "retry succeeded by secondary.", plugin_id: plugin_id, chunk_id: dump_unique_id_hex(chunk_id)
+              log.warn "retry succeeded by secondary.", chunk_id: dump_unique_id_hex(chunk_id)
```
Comment (repeatedly) [inline lib/fluent/plugin/output.rb]: Why remove plugin_id from the logs? Redundancy?

File: lib/fluent/plugin_helper/socket.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      # close    : [-]
+      # terminate: [-]
+
+      include Fluent::PluginHelper::SocketOption
+
+      attr_reader :_sockets # for tests
+
+      # TODO: implement connection pool for specified host
+
+      def socket_create(host, port, proto: :tcp, **kwargs, &block)
```
Comment (repeatedly) [inline lib/fluent/plugin_helper/socket.rb]: How about `proto` argument is changed to required?

File: lib/fluent/plugin_helper/socket.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        when :tcp
+          socket_create_tcp(host, port, **kwargs, &block)
+        when :udp
+          socket_create_udp(host, port, **kwargs, &block)
+        when :tls
+          socket_create_tls(host, port, **kwargs, &block)
+        when :unix
+          raise "not implemented yet"
+        else
+          raise ArgumentError, "invalid protocol"
```
Comment (repeatedly) [inline lib/fluent/plugin_helper/socket.rb]: "invalid protocol: #{proto}" is better.

File: lib/fluent/plugin/output.rb (hunk 1 above)
Comment (tagomoris) [inline lib/fluent/plugin/output.rb]: Redundancy. Plugin logger fills it automatically now.

File: lib/fluent/plugin/output.rb [hunk 4]
```diff
@@ -997,43 +1005,60 @@ def try_flush
...
+          update_retry_state(chunk.unique_id, using_secondary, e)
+
+          raise if @under_plugin_development && !@retry_for_error_chunk
+        end
+      end
+
+      def update_retry_state(chunk_id, using_secondary, error = nil)
+        @retry_mutex.synchronize do
+          @counters_monitor.synchronize{ @num_errors += 1 }
+          chunk_id_hex = dump_unique_id_hex(chunk_id)
```
Comment (tagomoris) [inline lib/fluent/plugin/output.rb]: There are almost no pass not to output a log with `chunk_id_hex`.
It's just:
* `@retry` exists, and
  * without error (called from `#rollback_commit`)
  * or
  * `@retry.limit?` is true (all chunks will be cleared)

That variable is referred in most cases, so I think it's worth to assign into a variable to reduce code complexity.

File: lib/fluent/plugin_helper/socket.rb (hunk 2 above)
Comment (tagomoris) [inline lib/fluent/plugin_helper/socket.rb]: Good idea. I'll update to `socket_create(proto, host, port, **kwargs, &block)`.

File: lib/fluent/plugin_helper/socket.rb (hunk 3 above)
Comment (tagomoris) [inline lib/fluent/plugin_helper/socket.rb]: gotcha.

**Pull request merged**